tests/api/ai_help_history.rs

```diff
@@ -95,6 +103,105 @@ async fn test_history() -> Result<(), Error> {
     Ok(())
 }
 
+#[actix_rt::test]
+#[stubr::mock(port = 4321)]
+async fn test_history_deletion() -> Result<(), Error> {
+    let pool = reset()?;
+    let app = test_app_with_login(&pool).await.unwrap();
+    let service = test::init_service(app).await;
+    let mut _logged_in_client = TestHttpClient::new(&service).await;
+    let mut conn = pool.get()?;
+
+    let history_deletion_period_in_sec = &SETTINGS
+        .ai
+        .as_ref()
+        .map(|ai| ai.history_deletion_period_in_sec)
+        .expect("ai.history_deletion_period_in_sec missing");
+
+    // Add an old chat history entry, double our configured period ago.
+    let ts = Utc::now()
+        .sub(Duration::from_secs(history_deletion_period_in_sec * 2))
+        .naive_utc();
+    let history = AIHelpHistoryInsert {
+        user_id: 1,
+        chat_id: Uuid::from_u128(1),
+        created_at: Some(ts),
+        updated_at: Some(ts),
+        label: "old entry".to_string(),
+    };
+    insert_into(ai_help_history::table)
+        .values(history)
+        .execute(&mut conn)?;
+
+    // Add a newer chat history entry, half of our configured period ago.
+    let ts = Utc::now()
+        // .checked_sub_months(Months::new(2))
+        .sub(Duration::from_secs(history_deletion_period_in_sec / 2))
+        .naive_utc();
+    let history = AIHelpHistoryInsert {
+        user_id: 1,
+        chat_id: Uuid::from_u128(2),
+        created_at: Some(ts),
+        updated_at: Some(ts),
+        label: "new entry".to_string(),
+    };
+    insert_into(ai_help_history::table)
+        .values(history)
+        .execute(&mut conn)?;
+
+    // check the history count before we run the delete job
+    let rec_count = ai_help_history::table
+        .filter(ai_help_history::user_id.eq(1))
+        .select(count(ai_help_history::user_id))
+        .first::<i64>(&mut conn)?;
+
+    assert_eq!(2, rec_count);
+
+    // Now, run the delete job.
+    let req = test::TestRequest::post()
+        .uri("/admin-api/ai-history/")
+        .insert_header((
+            "Authorization",
+            format!("Bearer {}", SETTINGS.auth.admin_update_bearer_token),
+        ))
+        .to_request();
+
+    let res = test::call_service(&service, req).await;
+    assert!(res.response().status().is_success());
+
+    // Check database that the old entry is gone.
+    // Loop until we see the old entry is gone because the
+    // delete job runs asynchonously.
+    let mut retry = 0;
+    const MAX_RETRIES: u32 = 40;
+    let mut records: Vec<NaiveDateTime>;
+    loop {
+        records = ai_help_history::table
+            .filter(ai_help_history::user_id.eq(1))
+            .select(ai_help_history::updated_at)
+            .get_results(&mut conn)?;
+        if records.len() == 1 {
+            break;
+        }
+
+        actix_rt::time::sleep(Duration::from_millis(5)).await;
+        retry += 1;
+        if retry > MAX_RETRIES {
+            break;
+        }
+    }
```

Comment:
Maybe this could be made more readable by using the [`retry` crate](https://docs.rs/retry/latest/retry/), and we could reuse it to stabilize those flaky tests?

Reply:
Yes, probably. Lets look into retry crate next time we have flakyness issues.